Add in-game staff channel

xenonca opened the pull request:
See https://github.com/MT-CTF/ctf_pvp_engine/pull/42

xenonca committed:
Add in-game staff channel

Moved from https://github.com/MT-CTF/ctf_pvp_engine/pull/42

ClobberXD commented on init.lua:
```diff
@@ -222,3 +222,24 @@ minetest.register_chatcommand("whereis", {
+            minetest.log("action", "CHAT[STAFFCHANNEL]: <" .. name .. "> " .. param)
+		if minetest.global_exists("chatplus") then
+			chatplus.log("[STAFFCHANNEL]: <" .. name .. "> " .. param)
+		end
```
Lines 239 to 241 can be removed. `chatplus` no longer exists, and it's a waste of effort to add stuff for something that doesn't exist. `ctf_pvp_engine` has quite a few `if chatplus then do something` sorts of code, but only because it was developed in a time when CTF had chatplus and so rubenwardy wanted it to support chatplus-specific features.

xenonca replied:
Done

xenonca committed:
Update init.lua

ClobberXD commented on init.lua:
```diff
@@ -222,3 +222,21 @@ minetest.register_chatcommand("whereis", {
+            table.insert(staff, toname)
+            minetest.chat_send_player(toname, minetest.colorize("#ff6600",
+                    "«" .. name .. "» " .. param))
+            minetest.log("action", "CHAT[STAFFCHANNEL]: <" .. name .. "> " .. param)
```
```suggestion
            minetest.log("action", "CHAT [STAFFCHANNEL]: <" .. name .. "> " .. param)
```

ClobberXD commented on init.lua:
```diff
@@ -222,3 +222,21 @@ minetest.register_chatcommand("whereis", {
+minetest.register_chatcommand("s", {
+    params = "<msg>",
+    description = "Send a message on the staff channel",
+    privs = {kick=true},
```
Spaces around `=` and within braces improve readability.

```suggestion
    privs = { kick = true },
```

ClobberXD commented on init.lua:
```diff
@@ -222,3 +222,21 @@ minetest.register_chatcommand("whereis", {
+    local staff = {}
+    for _, player in pairs(minetest.get_connected_players()) do
+        local toname = player:get_player_name()
+        if minetest.check_player_privs(toname, {kick=true}) then
```
It's better to also check for `ban`

```suggestion
        if minetest.check_player_privs(toname, { kick = true, ban = true }) then
```

ClobberXD replied:
Same for `privs`

ClobberXD reviewed (dismissed):
- Some simple suggestions in the following line comments...
- Please replace all whitespace indents with tab-space indents. This can be automated, but the process varies from IDE to IDE (or code editor to code editor).

****

Names of staff members are being added to a _local_ table `staff` every time the chat-command is invoked, which is unnecessary. Not to mention, the table is also unused. :P

A more performant implementation would be to:

- Declare `staff` as a local variable in file scope (i.e. outside `minetest.register_chatcommand`).
- Move priv check, and name insertion/removal to `on_joinplayer` and `on_leaveplayer` callbacks, like so:

  ```lua
  minetest.register_on_joinplayer(function(player)
  	if minetest.check_player_privs(player, { kick = true, ban = true }) then
  		table.insert(staff, player:get_player_name())
  	end
  end)

  minetest.register_on_leaveplayer(function(player)
  	local name = player:get_player_name()
  	local idx = table.indexof(staff, name)
  	if idx ~= -1 then 		-- table.indexof returns -1 if value doesn't exist in the table
  		table.remove(staff, idx)
  	end
  end)
  ```

  (If these callbacks already exist, then just append your code to the existing ones)

- In the chat-command's `func`, simply iterate over all names in `staff` like so:

  ```lua
  func = function(name, param)
  	for _, toname in pairs(staff) do
  		minetest.chat_send_player(toname, minetest.colorize("#ff6600",
  			"«" .. name .. "» " .. param))
  		minetest.log("action", "CHAT [STAFFCHANNEL]: <" .. name .. "> " .. param)
  	end
  end
  ```

This benefit of the suggested implementation is that the code doesn't iterate over all online players every single time this chat-command is invoked, which could potentially harm performance - we can easily keep track of all staff members using on_join/leaveplayer callbacks.

xenonca committed:
Update init.lua

xenonca committed:
Update init.lua

Change staff channel message color to better distinguish them from red team messages

xenonca commented:
Should work now but I think I need to open a new PR for this since my fork seems outdated

ClobberXD commented:
Updating the PR branch isn't required, unless there's a merge conflict.

ClobberXD commented on init.lua:
```diff
@@ -222,3 +222,31 @@ minetest.register_chatcommand("whereis", {
+minetest.register_chatcommand("s", {
+	params = "<msg>",
+	description = "Send a message on the staff channel",
+	privs = { ban = true },
```
Check for `kick` and `ban` here:

```suggestion
	privs = { kick = true, ban = true },
```

xenonca committed:
Update init.lua

Co-Authored-By: ANAND <ClobberXD@gmail.com>

Thomas--S commented on init.lua:
```diff
@@ -222,3 +222,31 @@ minetest.register_chatcommand("whereis", {
+	func = function(name, param)
+		for _, toname in pairs(staff) do
+			minetest.chat_send_player(toname, minetest.colorize("#ff9900",
+				"«" .. name .. "» " .. param))
```
I think it might be better to avoid non-ASCII signs here.

ClobberXD replied:
Agreed

Thomas--S reviewed (commented):
It might be worth considering to allow staff members to subscribe to this channel via IRC (similiar to how reports work).

Other than these comments I approve the concept, but I haven't tested it yet.

xenonca committed:
Replace  Non-ASCII Chars

xenonca commented:
> It might be worth considering to allow staff members to subscribe to this channel via IRC (similiar to how reports work).

What about just bridging it to the IRC staff channel? Is that possible without needing another bridge?

ClobberXD commented on init.lua:
```diff
@@ -231,7 +231,7 @@ minetest.register_chatcommand("s", {
 		for _, toname in pairs(staff) do
 			minetest.chat_send_player(toname, minetest.colorize("#ff9900",
-				"«" .. name .. "» " .. param))
+				"#" .. name .. ": " .. param))
```
Hmm, why have you used `#` and `:`?

xenonca replied:
``<<player>>`` seemed too long and confusing to me, but i can change it to that if you wish, or to anything else if you have a better idea

ClobberXD replied:
I think `<player> message` would suit this just fine, thanks to the message being colorized. Once CTF supports 5.1.0, we could then make use of the customisable chat message format (https://github.com/minetest/minetest/commit/d1c27c7e8037e9f438741ee5f7d3b6bda22d22a0).

xenonca committed:
Update init.lua

ClobberXD commented:
I completely forgot about this PR, sorry.
